Test authentication when Key Vault switches tenants

heaths opened the pull request:
Fixes #35086

azure-sdk commented:
**API change check**

API changes are not detected in this pull request.

heaths committed:
Support switching tenant IDs

Fixes #25086

heaths commented on sdk/resourcemanager/Azure.ResourceManager/src/Resources/Custom/TenantResource.cs:
```diff
@@ -16,8 +16,10 @@ namespace Azure.ResourceManager.Resources
     [CodeGenSuppress("GetAsync", typeof(CancellationToken))]
     [CodeGenSuppress("GetAvailableLocations", typeof(CancellationToken))]
     [CodeGenSuppress("GetAvailableLocationsAsync", typeof(CancellationToken))]
+#pragma warning disable CA1825 // Avoid zero-length array allocations
```
@m-nash this was necessary with Visual Studio 17.6.0 Preview 2.0 and has been for a little while, IIRC. We may need a constructor overload because you can't pass `Array.Empty<Type>()` since it's not a constant expression. Or is this a known issue?

annelo-msft replied:
I've seen this issue locally - is it possible to fix the issue rather than disable the warning?

annelo-msft replied:
Would it make sense to file an issue to track the work?

JoshLove-msft commented on sdk/core/Azure.Core/tests/BearerTokenAuthenticationPolicyTests.cs:
```diff
@@ -789,6 +789,109 @@ public async Task BearerTokenAuthenticationPolicy_BackgroundRefreshFailsAndLogs(
+            Assert.AreEqual("de763a21-49f7-4b08-a8e1-52c8fbc103b4", policy.TenantId);
+
+            await SendGetRequest(transport, policy, uri: new("https://example.com/1/Original"));
+            Assert.AreEqual("72f988bf-86f1-41af-91ab-2d7cd011db47", policy.TenantId);
```
Is there a way to validate that the new token is generated when the tenant changes?

heaths replied:
I can see, but at some point I'm just going to end up reimplementing Key Vault's `ChallengeBasedAuthenticationPolicy`, which already tests this scenario. Maybe I should just copy it and be done with it rather than try to come up with a minimal repro.

JoshLove-msft commented on sdk/core/Azure.Core/tests/BearerTokenAuthenticationPolicyTests.cs:
```diff
@@ -789,6 +789,109 @@ public async Task BearerTokenAuthenticationPolicy_BackgroundRefreshFailsAndLogs(
+            Assert.AreEqual("72f988bf-86f1-41af-91ab-2d7cd011db47", policy.TenantId);
+        }
+
+        private class ChallengeBasedAuthenticationPolicy : BearerTokenAuthenticationPolicy
```
nit: consider using Test in the name - I initially thought this was an released policy when I saw it used in the test.

christothes commented on sdk/core/Azure.Core/tests/BearerTokenAuthenticationPolicyTests.cs:
```diff
@@ -789,6 +789,109 @@ public async Task BearerTokenAuthenticationPolicy_BackgroundRefreshFailsAndLogs(
         }
 
+        [Test]
+        public async Task BearerTokenAuthenticationPolicy_SwitchedTenants()
```
This feels like it should be a test for the real `ChallengeBasedAuthenticationPolicy`. The unit test here could be simplified to just validate that the cache recognizes a change in tenantId.

heaths replied:
I already added one in #35474 but you said you wanted one local to Azure.Core.

christothes replied:
Ah, right. But I think the local one here should only need to validate the new token cache invalidation behavior.

heaths replied:
It should now with the latest push. Please have another look.

heaths committed:
Resolve PR feedback

christothes commented on sdk/core/Azure.Core/tests/BearerTokenAuthenticationPolicyTests.cs:
```diff
@@ -789,6 +789,114 @@ public async Task BearerTokenAuthenticationPolicy_BackgroundRefreshFailsAndLogs(
+            string tenantId = null;
+            var credential = new TokenCredentialStub((r, c) =>
+            {
+                tenantId = r.TenantId;
```
should we also validate that the stub only gets invoked twice to validate that the cache satisfies the request as expected?

heaths committed:
Assert the call count

Alternatively, I could increment the count only when the tenant ID is specified, but @christothes and I agree the comments here are clearer.

heaths commented:
I opened #36059 to track the failing test; otherwise, @JoshLove-msft and @christothes are good with the merge.

heaths commented:
/check-enforcer override
